File: code/backend/src/modules/auth/authService.ts
import bcrypt from "bcrypt";
import jwt from "jsonwebtoken";

import type { FastifyReply } from "fastify";

import { env } from "../../env.js";
import { AuthRepository } from "./authRepository.js";

const authRepository = new AuthRepository();

export class AuthService {
  async signup(
    data: {
      nome: string;
      email: string;
      senha: string;
      telefone?: string;

      role: "CLIENTE" | "ESTILISTA";

      especialidade?: string;
      descricao?: string;
    },
    reply: FastifyReply,
  ) {
    const existingUser = await authRepository.findUserByEmail(data.email);

    if (existingUser) {
      throw new Error("Email já cadastrado");
    }

    const senhaHash = await bcrypt.hash(data.senha, 10);

    const user = await authRepository.createUser({
      email: data.email,
      senha: senhaHash,
      role: data.role,
    });

    let clienteOuEstilista: any = null;

    if (data.role === "CLIENTE") {
      clienteOuEstilista = await authRepository.createCliente({
        nome: data.nome,
        ...(data.telefone && {
          telefone: data.telefone,
        }),
        userId: user.id,
      });
    }

    if (data.role === "ESTILISTA") {
      clienteOuEstilista = await authRepository.createEstilista({
        nome: data.nome,

        ...(data.telefone && {
          telefone: data.telefone,
        }),

        especialidade: data.especialidade ?? "",

        ...(data.descricao && {
          descricao: data.descricao,
        }),

        userId: user.id,
      });
    }

    const token = jwt.sign(
      {
        userId: user.id,
        role: user.role,
      },
      env.JWT_SECRET,
      {
        expiresIn: "7d",
      },
    );

    reply.setCookie("token", token, {
      httpOnly: true,
      sameSite: "lax",
      secure: false,
      path: "/",
    });

    return {
      message: "Usuário criado com sucesso",
      token,
      user: {
        id: user.id,
        email: user.email,
        role: user.role,
      },
      profile: clienteOuEstilista,
    };
  }

  async login(
    data: {
      email: string;
      senha: string;
    },
    reply: FastifyReply,
  ) {
    const user = await authRepository.findUserByEmail(data.email);

    if (!user) {
      throw new Error("Email ou senha inválidos");
    }

    const senhaCorreta = await bcrypt.compare(data.senha, user.senha);

    if (!senhaCorreta) {
      throw new Error("Email ou senha inválidos");
    }

    const token = jwt.sign(
      {
        userId: user.id,
        role: user.role,
      },
      env.JWT_SECRET,
      {
        expiresIn: "7d",
      },
    );

    reply.setCookie("token", token, {
      httpOnly: true,
      sameSite: "lax",
      secure: false,
      path: "/",
    });

    // Buscar dados do cliente ou estilista
    let profile = null;
    if (user.role === "CLIENTE") {
      profile = await authRepository.findClienteByUserId(user.id);
    } else if (user.role === "ESTILISTA") {
      profile = await authRepository.findEstilistaByUserId(user.id);
    }

    return {
      message: "Login realizado com sucesso",
      token,
      user: {
        id: user.id,
        email: user.email,
        role: user.role,
      },
      profile,
    };
  }

  async logout(reply: FastifyReply) {
    reply.clearCookie("token");

    return {
      message: "Logout realizado com sucesso",
    };
  }
}

File: code/mist_mobile/lib/services/auth_service.dart
import 'api_client.dart';
import 'cookie_storage.dart';

class AuthService {
  AuthService({ApiClient? apiClient, CookieStorage? cookieStorage})
      : _apiClient = apiClient ?? ApiClient(),
        _cookieStorage = cookieStorage ?? CookieStorage();

  final ApiClient _apiClient;
  final CookieStorage _cookieStorage;

  Future<AuthSession> login({
    required String email,
    required String senha,
  }) async {
    final data = await _apiClient.post(
      '/auth/login',
      body: {
        'email': email,
        'senha': senha,
      },
    ) as Map<String, dynamic>;

    return AuthSession.fromJson(data);
  }

  Future<AuthSession> signup({
    required String nome,
    required String email,
    required String senha,
    required String role,
    String? telefone,
    String? especialidade,
    String? descricao,
  }) async {
    final body = <String, dynamic>{
      'nome': nome,
      'email': email,
      'senha': senha,
      'role': role,
      if (telefone != null && telefone.isNotEmpty) 'telefone': telefone,
      if (especialidade != null && especialidade.isNotEmpty)
        'especialidade': especialidade,
      if (descricao != null && descricao.isNotEmpty) 'descricao': descricao,
    };

    final data = await _apiClient.post('/auth/signup', body: body)
        as Map<String, dynamic>;

    return AuthSession.fromJson(data);
  }

  Future<void> logout() async {
    try {
      await _apiClient.post('/auth/logout');
    } finally {
      await _cookieStorage.clear();
    }
  }

  Future<AuthSession?> checkAuth() async {
    try {
      final data = await _apiClient.get('/auth/check-auth') as Map<String, dynamic>;
      return AuthSession.fromJson(data);
    } on ApiException {
      await _cookieStorage.clear();
      return null;
    }
  }
}

class AuthSession {
  const AuthSession({
    required this.authenticated,
    required this.userId,
    required this.role,
    this.email,
    this.profile,
  });

  final bool authenticated;
  final int userId;
  final String role;
  final String? email;
  final Map<String, dynamic>? profile;

  factory AuthSession.fromJson(Map<String, dynamic> json) {
    final user = json['user'] as Map<String, dynamic>? ?? {};

    return AuthSession(
      authenticated: json['authenticated'] as bool? ?? true,
      userId: user['id'] as int? ?? 0,
      role: user['role']?.toString() ?? '',
      email: user['email']?.toString(),
      profile: json['profile'] as Map<String, dynamic>?,
    );
  }
}
